kernel/src/expressions/column_names.rs

```diff
@@ -106,6 +119,190 @@ impl Hash for ColumnName {
+/// ```
+///
+/// The string representation is lossless, and can be parsed back into a `ColumnName` using
+/// [`FromStr`]:
+///
+/// ```
+/// # use delta_kernel::expressions::ColumnName;
+/// let colname = ColumnName::new(["a", "b.c", "d"]);
+/// let parsed: ColumnName = colname.to_string().parse().unwrap();
+/// assert_eq!(colname, parsed);
+/// ```
+impl Display for ColumnName {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut delim = None;
+        for s in self.iter() {
+            use std::fmt::Write as _;
+
+            // Only emit the period delimiter after the first iteration
+            if let Some(d) = delim {
+                f.write_char(d)?;
+            } else {
+                delim = Some('.');
+            }
+
+            let digit_char = |c: char| c.is_ascii_digit();
+            let special_char = |c: char| !c.is_ascii_alphanumeric() && c != '_';
+            if s.is_empty() || s.starts_with(digit_char) || s.contains(special_char) {
+                // Special situation detected. For safety, surround the field name with backticks
+                // (with proper escaping if the field name itself contains backticks).
+                f.write_char('`')?;
+                for c in s.chars() {
+                    match c {
+                        '`' => f.write_str("``")?,
+                        _ => f.write_char(c)?,
+                    }
+                }
+                f.write_char('`')?;
+            } else {
+                // The fild name contains no special characters, so emit it as-is.
+                f.write_str(s)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+fn drop_leading_whitespace(iter: &mut Peekable<impl Iterator<Item = char>>) {
+    while iter.next_if(|c| c.is_whitespace()).is_some() {}
+}
+
+/// Parses a column name from a string. Field names are separated by dots. Whitespace between fields
+/// is ignored. Field names enclosed in backticks may contain arbitrary characters, including
+/// periods and spaces. To include a literal backtick in a field name, escape it by doubling, e.g.:
+///
+/// ```
+/// # use delta_kernel::expressions::ColumnName;
+/// assert_eq!(ColumnName::new(["a", "b.`c`.d", "e"]).to_string(), "a.`b.``c``.d`.e");
+/// ```
+///
+/// NOTE: Unlike the conversion from `ColumnName` to `String` and back, a conversion from `String`
+/// to `ColumnName` and back may not exactly match the original string, if the latter included
+/// whitespace or unnecessary field escapes, e.g.:
+///
+/// ```
+/// # use delta_kernel::expressions::ColumnName;
+/// let parsed: ColumnName = " `a` . `b.``c``.d` . `e` ".parse().unwrap();
+/// assert_eq!(parsed.to_string(), "a.`b.``c``.d`.e");
+/// ```
+impl std::str::FromStr for ColumnName {
+    type Err = Error;
+
+    fn from_str(s: &str) -> DeltaResult<Self> {
+        // Ambiguous case: The empty string `""`could reasonably parse as either `ColumnName::new([""])`
```

Comment:
nit, i think this comment belongs two lines lower

Reply:
Will fix in the follow-up PR that adds support for parsing column lists, since this code will anyway move.